torch/lib/c10d/ProcessGroupGloo.hpp

```diff
@@ -0,0 +1,184 @@
+  std::vector<std::vector<int64_t>> dstSizes;
+  int srcDevice = -1;
+  int dstDevice = -1;
+  int srcRank = -1;
+  int dstRank = -1;
+  ReduceOp reduceOp = ReduceOp::UNUSED;
+};
+
+struct AlgorithmEntry {
+  AlgorithmKey key;
+  std::unique_ptr<::gloo::Algorithm> algorithm;
+  std::vector<at::Tensor> src;
+  std::vector<at::Tensor> dst;
+  std::function<void(std::unique_ptr<AlgorithmEntry>&)> run;
+
+  explicit AlgorithmEntry() {
+  }
+
+  // Must not be copied
+  AlgorithmEntry & operator=(const AlgorithmEntry&) = delete;
+  AlgorithmEntry(const AlgorithmEntry&) = delete;
+
+  // May be moved
+  AlgorithmEntry(AlgorithmEntry&& other) {
+    key = std::move(other.key);
+    algorithm = std::move(other.algorithm);
+    src = std::move(other.src);
+    dst = std::move(other.dst);
+  }
+};
+
+} // namespace c10d
+
+MAKE_HASHABLE(
+    ::c10d::AlgorithmKey,
+    t.collectiveType,
+    t.type,
+    t.devices,
+    t.srcSizes,
+    t.dstSizes,
+    t.srcDevice,
+    t.dstDevice,
+    t.srcRank,
+    t.dstRank,
+    t.reduceOp);
+
+namespace c10d {
+
+class ProcessGroupGloo : public ProcessGroup {
+ public:
+  class WorkGloo : public Work {
+   public:
+    explicit WorkGloo();
+    virtual ~WorkGloo();
+
+    // Checks if request has completed. Non-blocking operation.
+    bool isCompleted() override;
+
+    // Waits until request completes. Blocking operation.
+    // Returns false if the work completed with an exception.
+    bool wait() override;
+
+    // Returns exception if wait() returned false.
+    const std::exception& exception() const override;
+
+   protected:
+    void finish();
+    void finishWithException(const ::gloo::Exception& ex);
+
+    std::mutex m_;
+    std::condition_variable cv_;
+    bool completed_;
+
+    // Use pointer to ::gloo::Exception because it doesn't have a
+    // default constructor and constructing an empty std::unique_ptr
+    // is probably cheaper (this is highly speculative).
+    std::unique_ptr<::gloo::Exception> ex_;
+
+    friend class ProcessGroupGloo;
+  };
+
+  explicit ProcessGroupGloo(
+      const std::shared_ptr<Store>& store,
+      const std::vector<std::shared_ptr<::gloo::transport::Device>>& devices,
+      int rank,
+      int size);
+
+  virtual ~ProcessGroupGloo();
+
+  void initialize();
+  void destroy();
+
+  std::shared_ptr<Work> broadcast(
+      std::vector<at::Tensor>& data,
+      const BroadcastOptions& opts = BroadcastOptions()) override;
+
+  std::shared_ptr<Work> allreduce(
+      std::vector<at::Tensor>& tensors,
+      const AllreduceOptions& opts = AllreduceOptions()) override;
+
+ protected:
+  std::unique_ptr<::gloo::rendezvous::Store> store_;
+  std::vector<std::shared_ptr<::gloo::transport::Device>> devices_;
+  std::vector<std::shared_ptr<::gloo::Context>> contexts_;
+
+  using KeyType = AlgorithmKey;
+  using EntryType = std::unique_ptr<AlgorithmEntry>;
+
+  EntryType construct(const KeyType& key);
+  EntryType checkout(const KeyType& key);
+
+  std::mutex m_;
```

Comment:
consider renaming this mutex, we have two mutex in two classes with the same name m_?

Reply:
Suggestions? I like m_ for being an instance wide mutex, for whatever instance you're dealing with.